### PR TITLE
Spring Boot 2.0+ support

### PR DESCRIPTION
**Pull request by binakot**
Init implementation for Spring Boot 2.0+ support. It's not backward compatible with Spring Boot 1.0+ obviously. https://github.com/mmnaseri/spring-data-mock/issues/163

The main differences:
0) Switch to Java 8, update all dependencies to the last versions.
1) CrudRepository and Auditable return Optional of value instead of value itself.
2) The key in CrudRepository doesn't implement Serializable interface anymore.
3) There are a lot of changes from constructor creating with `new` keyword to static builder methods.
4) DateTime from JodaTime was changed to Instant from DateTime API. JodaTime was removed from dependencies at all.

### CODE REVIEW

File: spring-data-mock/src/main/java/com/mmnaseri/utils/spring/data/domain/impl/AssignableRepositoryMetadataResolver.java [hunk 1]
```diff
@@ -25,7 +23,7 @@ protected RepositoryMetadata resolveFromInterface(Class<?> repositoryInterface)
         final Class<?>[] arguments = GenericTypeResolver.resolveTypeArguments(repositoryInterface, Repository.class);
         final Class<?> entityType = arguments[0];
         final Class<?> rawIdType = arguments[1];
-        final Class<? extends Serializable> idType = rawIdType.asSubclass(Serializable.class);
+        final Class<?> idType = rawIdType.asSubclass(Object.class);
         final String idProperty = resolveIdProperty(entityType, idType);
         return new ImmutableRepositoryMetadata(idType, entityType, repositoryInterface, idProperty);
```
Comment (mmnaseri) [inline spring-data-mock/src/main/java/com/mmnaseri/utils/spring/data/domain/impl/AssignableRepositoryMetadataResolver.java:28]:
```suggestion
        final String idProperty = resolveIdProperty(entityType, rawIdType);
        return new ImmutableRepositoryMetadata(rawIdType, entityType, repositoryInterface, idProperty);
```

File: spring-data-mock/src/main/java/com/mmnaseri/utils/spring/data/dsl/mock/KeyGeneratorProvider.java [hunk 2]
```diff
@@ -22,15 +27,15 @@
...
         final List<Class<? extends KeyGenerator>> discoveredKeyGenerators = getKeyGeneratorTypes();
         generators = new ConcurrentHashMap<>();
         for (Class<? extends KeyGenerator> generatorType : discoveredKeyGenerators) {
             final Class<?> keyType = GenericTypeResolver.resolveTypeArgument(generatorType, KeyGenerator.class);
             assert keyType != null;
-            final Class<? extends Serializable> actualKeyType = keyType.asSubclass(Serializable.class);
+            final Class<?> actualKeyType = keyType.asSubclass(Object.class);
             if (!generators.containsKey(actualKeyType)) {
                 generators.put(actualKeyType, new CopyOnWriteArrayList<Class<? extends KeyGenerator>>());
             }
```
Comment (mmnaseri) [inline spring-data-mock/src/main/java/com/mmnaseri/utils/spring/data/dsl/mock/KeyGeneratorProvider.java:41]:
```suggestion
            if (!generators.containsKey(keyType)) {
                generators.put(keyType, new CopyOnWriteArrayList<Class<? extends KeyGenerator>>());
            }
```
Also, the line after this, but Github doesn't let me comment on that 🤷‍♂️

File: spring-data-mock/src/main/java/com/mmnaseri/utils/spring/data/store/impl/AuditDataEventListener.java [hunk 3]
```diff
@@ -3,10 +3,11 @@
 import com.mmnaseri.utils.spring.data.domain.RepositoryMetadata;
 import com.mmnaseri.utils.spring.data.store.DataStoreEvent;
 import com.mmnaseri.utils.spring.data.store.DataStoreEventListener;
-import org.joda.time.DateTime;
 import org.springframework.data.domain.Auditable;
 import org.springframework.data.domain.AuditorAware;
 
+import java.time.Instant;
```
Comment (mmnaseri) [inline spring-data-mock/src/main/java/com/mmnaseri/utils/spring/data/store/impl/AuditDataEventListener.java:9]: Does this mean that jodatime can be removed as a dependency?

**Comment by mmnaseri** [comment]
@binakot Thanks a lot for taking care of this. I know it took me a couple of days to go through it all, but I finally did it. If you can make the couple of tiny, nit-picking changes I have requested, I'll merge. Otherwise, if you think you don't have the time for those, I can merge and do it myself.

Also, I think these are breaking changes, that need to start a new major version to reflect the change from Spring Data 1.x to 2.x. What do you think?

File: spring-data-mock/src/main/java/com/mmnaseri/utils/spring/data/store/impl/AuditDataEventListener.java (hunk 3 above)
Comment (binakot) [reply spring-data-mock/src/main/java/com/mmnaseri/utils/spring/data/store/impl/AuditDataEventListener.java:9]: It's already removed :)

**Comment by binakot** [comment]
@mmnaseri Great! I'll do changes myself, no problems :)
Yes, it's not backward compatible and will work with spring boot 2+ only. New major is true way.
And I think you should update readme after new release to let to know developers about it :octocat:

**Pull request merged**